### PR TITLE
[red-knot] Refactoring the inference logic of lexicographic comparisons

### LINKED ISSUE #14279: [red-knot] `Literal[False]` not always inferred for comparisons between tuple types with disjunct literal elements
red-knot infers the following types for the following comparisons:

```py
from typing import reveal_type

def int_instance() -> int:
    return 42

reveal_type("foo" == "bar")                                       # Literal[False]
reveal_type(("foo",) == ("bar",))                                 # Literal[False]
reveal_type((4, "foo") == (4, "bar"))                             # Literal[False]
reveal_type((int_instance(), "foo") == (int_instance(), "bar"))   # bool
```

The last one should also be inferred as `Literal[False]`, just like the first three. This is because two objects of types `Literal["foo"]` and `Literal["bar"]` respectively will never compare equal to each other. By the same token, two 2-element tuples where the second elements are of type `Literal["foo"]` and `Literal["bar"]` respectively will never compare equal, regardless of what their first element is.

I discovered this while working on https://github.com/astral-sh/ruff/pull/14244. Ideally, this comparison would be inferred as `Literal[False]`, since `sys.version_info[3]` should be inferred by red-knot as being of type `Literal["alpha", "beta", "candidate", "final"]`, and objects with that type never compare equal to the string `"finallllllll"`:

```py
import sys
from typing import reveal_type

reveal_type(sys.version_info == (3, 8, 1, "finallllllll", 0))
```

### PR DESCRIPTION
**Pull request by cake-monotone**
## Summary

closes #14279

#### Incorrect Error Propagation

In the current implementation of lexicographic comparisons, if the result of an Eq operation is Ambiguous, the comparison stops immediately, returning a bool instance. While this may yield correct inferences, it fails to capture unsupported-operation errors that might occur in subsequent comparisons.
```py
class A: ...

(int_instance(), A()) < (int_instance(), A())  # should error
```

#### Weak Inference in Specific Cases

> Example: `(int_instance(), "foo") == (int_instance(), "bar")`
> Current result: `bool`
> Expected result: `Literal[False]`

`Eq` and `NotEq` have unique behavior in lexicographic comparisons compared to other operators. Specifically:
- For `Eq`, if any non-equal pair exists within the tuples being compared, we can immediately conclude that the tuples are not equal.
- For `NotEq`, if any equal pair exists, we can conclude that the tuples are unequal.

```py
a = (str_instance(), int_instance(), "foo")

reveal_type(a == a)  # revealed: bool
reveal_type(a != a)  # revealed: bool

b = (str_instance(), int_instance(), "bar")

reveal_type(a == b)  # revealed: bool  # should be Literal[False]
reveal_type(a != b)  # revealed: bool  # should be Literal[True]
```
#### Incorrect Support for Non-Boolean Rich Comparisons

In CPython, aside from `==` and `!=`, tuple comparisons return a non-boolean result as-is. Tuples do not convert the value into `bool`.

Note: If all pairwise `==` comparisons between elements in the tuples return Truthy, the comparison then considers the tuples' lengths. Regardless of the return type of the dunder methods, the final result can still be a boolean.

```py
from __future__ import annotations

class A:
    def __eq__(self, o: object) -> str:
        return "hello"

    def __ne__(self, o: object) -> bytes:
        return b"world"

    def __lt__(self, o: A) -> float:
        return 3.14

a = (A(), A())

reveal_type(a == a)  # revealed: bool
reveal_type(a != a)  # revealed: bool
reveal_type(a < a)  # revealed: bool # should be: `float | Literal[False]`

```

### Key Changes
One of the major changes is that comparisons no longer end with a `bool` result when a pairwise `Eq` result is `Ambiguous`. Instead, the function attempts to infer all possible cases and unions the results. This improvement allows for more robust type inference and better error detection.

Additionally, as the function is now optimized for tuple comparisons, the name has been changed from the more general `infer_lexicographic_comparison` to `infer_tuple_rich_comparison`.

## Test Plan

mdtest included

### CODE REVIEW

File: crates/red_knot_python_semantic/resources/mdtest/comparison/tuples.md [hunk 1]
```diff
@@ -134,23 +136,164 @@ reveal_type(c >= c)  # revealed: Literal[True]
...
+reveal_type((A(), B()) < (A(), B()))  # revealed: float | set | Literal[False]
+```
+
+#### Special Handling of Eq and NotEq in Lexicographic Comparisons
+
+> Example: `(int_instance(), "foo") == (int_instance(), "bar")`
+
+`Eq` and `NotEq` have unique behavior compared to other operators in lexicographic comparisons.
+Specifically, for `Eq`, if any non-equal pair exists within the tuples being compared, we can
+immediately conclude that the tuples are not equal. Conversely, for `NotEq`, if any equal pair
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/comparison/tuples.md:204]: Is this supposed to say "if any non-equal pair exists"?

File: crates/red_knot_python_semantic/resources/mdtest/comparison/tuples.md [hunk 2]
```diff
@@ -134,23 +136,164 @@ reveal_type(c >= c)  # revealed: Literal[True]
...
+exists, we can determine that the tuples are unequal.
+
+In contrast, with operators like `<` and `>`, the comparison must consider each pair of elements
+sequentially, and the final outcome might remain ambiguous until all pairs are compared.
+
+Naive implementations of lexicographic comparisons may overlook this distinction, leading to broad
+inferences.
+
+This is based on an issue reported [here](https://github.com/astral-sh/ruff/issues/14279).
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/comparison/tuples.md:213]: I don't think these paragraphs add value to the tests.

**Review by carljm — APPROVED** [review]
Excellent! Very clear write-up and fix, thank you!!

File: crates/red_knot_python_semantic/resources/mdtest/comparison/tuples.md (hunk 1 above)
Comment (cake-monotone) [reply crates/red_knot_python_semantic/resources/mdtest/comparison/tuples.md]: You're right, my mistake! 😅 Thanks for catching that.

**Comment by carljm** [comment]
Thank you!! So great.

**Pull request merged**